src/Symfony/Component/Form/DependencyInjection/FormPass.php

```diff
@@ -78,18 +85,20 @@ public function process(ContainerBuilder $container)
             }
 
             $typeExtensions[$extendedType][] = $serviceId;
+
+            // Add form type extension service to the service locator
+            $servicesMap[$serviceId] = new Reference($serviceId);
         }
 
         $definition->replaceArgument(2, $typeExtensions);
 
         $guessers = array_keys($container->findTaggedServiceIds($this->formTypeGuesserTag));
         foreach ($guessers as $serviceId) {
-            $serviceDefinition = $container->getDefinition($serviceId);
-            if (!$serviceDefinition->isPublic()) {
-                throw new InvalidArgumentException(sprintf('The service "%s" must be public as form type guessers are lazy-loaded.', $serviceId));
-            }
+            // Add form type guesser service to the service locator
+            $servicesMap[$serviceId] = new Reference($serviceId);
         }
 
+        $definition->replaceArgument(0, new ServiceLocatorArgument($servicesMap));
```

Comment:
Why not 2 service locators, and one iterator?

So the `DependencyInjectionExtension::__construct()` method signature would be:

```php
public function __construct(ContainerInterface $typesLocator, ContainerInterface $typeExtensionsLocator, iterator $guessers)
```

Might not be the easiest for BC though.